### PR TITLE
Support for auto-accessor fields from the Stage 3 Decorators proposal

### PR DESCRIPTION
**Pull request by rbuckton**
This PR adds support for `accessor` field declarations described in the [Stage 3 Decorators proposal](https://github.com/tc39/proposal-decorators).

An _Auto-Accessor_ is a field declaration that will be transformed by the runtime into a pair of `get` and `set` accessors that access a private backing field:

```js
class C {
  accessor x = 1;
}
// is transformed into
class C {
  // actual field is named `#x accessor storage` so that it isn't user-reachable
  #x_accessor_storage = 1;
  get x() { return this.#x_accessor_storage; }
  set x(value) { this.#x_accessor_storage = value; }
}
```

_Auto-Accessor_ fields have several capabilities:
- Allows subclasses to override the `get`/`set` without a superclass field potentially shadowing the property during initialization.
- A decorator applied to an auto-accessor receives the `get` and `set` accessor pair and can replace them without changing the runtime shape of the class.
- _Native_ ECMAScript decorators will be able to replace the initializer.

In addition, there are several rules around the use of the `accessor` keyword:
- `accessor` may only appear in front of field declarations on a `class`. It is not supported in `interface` or object type literals.
- `accessor` cannot be used with `readonly` or `declare` on the same field declaration.
- `accessor` *can* be used in an ambient class declaration.
- `accessor` field declarations can be decorated with TypeScript's legacy decorators (i.e., under `--experimentalDecorators`). They will behave as if you decorated a `get` or `set` declaration (i.e., you will receive a `PropertyDescriptor` at runtime with both `get` and `set` functions.
- A field explicitly marked with `accessor` will also be marked with `accessor` in the output declaration file.
- An `accessor` field does *not* make a class nominal, despite the synthetic private backing field.
- The TypeScript `Symbol` for an `accessor` field is not a `SymbolFlags.Property`, but rather a `SymbolFlags.GetAccessor | SymbolFlags.SetAccessor`.

NOTE: This is not an implementation of the full Stage 3 Decorators proposal as that effort is still in progress.

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by rbuckton** [comment]
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
> Why is this a thing? Like, what problem does it actually solve (genuine question, not snark)? The generated accessors don't do anything and unlike, e.g., C#, changing a plain property to an accessor isn't generally an API break...
> 
> _edit:_ Oh, I see, it's mostly because decorators will apparently make a distinction between field and accessor.

This was a requirement from VM implementers to avoid the potential hidden shape changes that could occur if you were able to convert a field declaration into an accessor dynamically rather than statically. It also helps to avoid issues with public field shadowing in cases like the following, such as with `@observe` in #48814:

```js
class C {
  // observe attaches an accessor pair to C.prototype, but the field `x` is defined
  // on the instance, shadowing the accessor pair.
  @observe x = 1;
}

// vs
class C {
  // observe can now instead intercept the `{ get, set }` provided to attach its behavior.
  @observe accessor x = 1;
}
```

**Comment by sandersn** [comment]
What are the assignability rules for auto-accessor fields? I guess that they're the same as a normal class field, but might be missing something.

**Comment by rbuckton** [comment]
They *should* be treated the same as we would a get/set pair.

**Comment by sandersn** [comment]
Ah, that explains why the code in the checker is with the existing accessor code, not the property declaration code.

**Review by sandersn — COMMENTED** [review]
Some initial questions about tests. I still need to read the emit, factory and transformer code though -- so the bulk of it.

File: tests/cases/conformance/decorators/class/property/decoratorOnClassProperty13.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// @target: ES2015
+// @experimentaldecorators: true
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/cases/conformance/decorators/class/property/decoratorOnClassProperty13.ts:2]: Yes. When Stage-3 decorators is implemented, you will only get the TS/legacy decorator emit using `--experimentalDecorators`.

File: tests/baselines/reference/accessorFieldDisallowedModifiers.errors.txt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+tests/cases/conformance/classes/propertyMemberDeclarations/accessorFieldDisallowedModifiers.ts(35,1): error TS1275: 'accessor' modifier can only appear on a property declaration.
+tests/cases/conformance/classes/propertyMemberDeclarations/accessorFieldDisallowedModifiers.ts(36,1): error TS1275: 'accessor' modifier can only appear on a property declaration.
+
+
+==== tests/cases/conformance/classes/propertyMemberDeclarations/accessorFieldDisallowedModifiers.ts (28 errors) ====
+    abstract class C1 {
+        accessor accessor a: any;
+                 ~~~~~~~~
+!!! error TS1030: 'accessor' modifier already seen.
+        readonly accessor b: any;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/baselines/reference/accessorFieldDisallowedModifiers.errors.txt:36]: `accessor` fields always generate a getter and a setter and we don't allow `readonly` on `get`/`set` declarations today. I'm also concerned that users might expect a `readonly accessor` to be `get`-only, but `readonly` has no effect on runtime behavior. The [Grouped and Auto-Accessors Proposal](https://github.com/tc39/proposal-grouped-and-auto-accessors) hopes to introduce `accessor x { get; }` as the mechanism for creating a `get`-only auto-accessor, so I'm more inclined to wait until that proposal advances. Unfortunately, that proposal is currently blocked until after Decorators hits Stage 4 and implementers can get feedback from the community on missing areas of coverage in Decorators.

File: tests/baselines/reference/accessorField5(target=es2015).js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+class C1 {
+    accessor ["w"]: any;
+    accessor ["x"] = 1;
+    static accessor ["y"]: any;
+    static accessor ["z"] = 2;
+}
+
+declare var f: any;
+class C2 {
+    // @ts-ignore
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/baselines/reference/accessorField5(target=es2015).js:11]: `A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.`. This is legal JavaScript however, we just issue an error since we can't reliably determine the shape of the class.

File: tests/baselines/reference/accessorField5(target=es2015).js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
+    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
+};
+var _a, _C1__a_accessor_storage, _C1__b_accessor_storage, _C1__c_accessor_storage, _C1__d_accessor_storage, _C2__e_accessor_storage, _b;
+class C1 {
+    constructor() {
+        _C1__a_accessor_storage.set(this, void 0);
+        _C1__b_accessor_storage.set(this, 1);
+    }
+    get [(_C1__a_accessor_storage = new WeakMap(), _C1__b_accessor_storage = new WeakMap(), "w")]() { return __classPrivateFieldGet(this, _C1__a_accessor_storage, "f"); }
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/baselines/reference/accessorField5(target=es2015).js:33]: Its primarily to support evaluation-order guarantees. We often do the same for *any* expressions that are evaluated inside the class body prior to evaluation of the expression in a computed property name, pretty much anywhere you see something like a `pendingExpressions` array in any of the transformers.

File: src/compiler/types.ts [hunk 5]
```diff
@@ -1310,8 +1319,16 @@ namespace ts {
         // escaping not strictly necessary
         // avoids gotchas in transforms and utils
         readonly escapedText: __String;
+        /*@internal*/ readonly autoGenerateFlags?: GeneratedIdentifierFlags; // Specifies whether to auto-generate the text for an identifier.
+        /*@internal*/ readonly autoGenerateId?: number;           // Ensures unique generated identifiers get unique names, but clones get the same name.
+        /*@internal*/ readonly autoGeneratePrefix?: string | GeneratedNamePart;
+        /*@internal*/ readonly autoGenerateSuffix?: string;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/types.ts:1325]: Auto-generation is entirely handled by the emitter, though I did add a change in this PR to also make the `.escapedText` of generated identifiers easier to recognize when debugging, though that text isn't actually used during emit.

File: src/compiler/types.ts [hunk 6]
```diff
@@ -1859,7 +1880,7 @@ namespace ts {
 
     export interface StringLiteral extends LiteralExpression, Declaration {
         readonly kind: SyntaxKind.StringLiteral;
-        /* @internal */ readonly textSourceNode?: Identifier | StringLiteralLike | NumericLiteral; // Allows a StringLiteral to get its text from another node (used by transforms).
+        /* @internal */ readonly textSourceNode?: Identifier | StringLiteralLike | NumericLiteral | PrivateIdentifier; // Allows a StringLiteral to get its text from another node (used by transforms).
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/types.ts:1883]: No, `textSourceNode` is used when we need to create a string literal that points to another identifier or literal. Usually when we synthesize a string for `Object.defineProperty`, but I will also be using this as part of the Decorators Stage-3 emit, which needs to emit a string literal for a private identifier name as part of the context object passed to a decorator. Most of the new name generation logic here was pulled directly from the branch where I'm working on the decorators transform.

**Comment by rbuckton** [comment]
@sandersn can you take another look?

**Review by sandersn — COMMENTED** [review]
Still need to look at classFields.ts and tests

File: tests/baselines/reference/autoAccessor5(target=es2015).js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+class C1 {
+    accessor ["w"]: any;
+    accessor ["x"] = 1;
+    static accessor ["y"]: any;
+    static accessor ["z"] = 2;
+}
+
+declare var f: any;
+class C2 {
+    // @ts-ignore
```
Comment (sandersn) [inline tests/baselines/reference/autoAccessor5(target=es2015).js:11]: what is the error? Do we not baseline emit when there are errors? I thought we did, but maybe I was thinking of types/symbols.

File: tests/baselines/reference/autoAccessor6(target=esnext,usedefineforclassfields=false).js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+class C3 extends C1 {
+    get a() { return super.a; }
+}
+
+
+//// [autoAccessor6.js]
+class C1 {
+}
```
Comment (sandersn) [inline tests/baselines/reference/autoAccessor6(target=esnext,usedefineforclassfields=false).js:17]: this emit is weird, but I *guess* correct?

File: tests/baselines/reference/autoAccessor7(target=es2022,usedefineforclassfields=false).js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+
+//// [autoAccessor7.js]
+class C1 {
+}
+class C2 extends C1 {
+    constructor() {
+        super(...arguments);
+        this.#a_accessor_storage = 1;
```
Comment (sandersn) [inline tests/baselines/reference/autoAccessor7(target=es2022,usedefineforclassfields=false).js:21]: does the emit of `#a_accessor_storage` matter in this case? Could it be the same as the emit for usedefineforclassfields=true?

File: tests/baselines/reference/autoAccessorExperimentalDecorators(target=esnext).errors.txt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        accessor a: any;
+    
+        @dec
+        static accessor b: any;
+    }
+    
+    class C2 {
+        @dec
+        ~
+!!! error TS1206: Decorators are not valid here.
```
Comment (sandersn) [inline tests/baselines/reference/autoAccessorExperimentalDecorators(target=esnext).errors.txt:19]: will ES decorators be valid here

File: tests/baselines/reference/privateNameDuplicateField.js [hunk 11]
```diff
@@ -557,7 +551,7 @@ function Method() {
     _d = A_Method_StaticSetter, _A_Method_StaticSetter_instances = new WeakSet();
 }
 function Getter() {
-    var _A_Getter_Field_instances, _A_Getter_Field_foo_get, _A_Getter_Field_foo, _A_Getter_Method_instances, _A_Getter_Method_foo_get, _A_Getter_Method_foo, _A_Getter_Getter_instances, _A_Getter_Getter_foo_get, _A_Getter_Getter_foo_get_1, _A_Getter_Setter_instances, _A_Getter_Setter_foo_get, _A_Getter_Setter_foo_set, _A_Getter_StaticField_instances, _a, _A_Getter_StaticField_foo_get, _A_Getter_StaticField_foo, _A_Getter_StaticMethod_instances, _b, _A_Getter_StaticMethod_foo_get, _A_Getter_StaticMethod_foo, _A_Getter_StaticGetter_instances, _c, _A_Getter_StaticGetter_foo_get, _A_Getter_StaticGetter_foo_get_1, _A_Getter_StaticSetter_instances, _d, _A_Getter_StaticSetter_foo_get, _A_Getter_StaticSetter_foo_set;
```
Comment (sandersn) [inline tests/baselines/reference/privateNameDuplicateField.js:560]: is this rearrangement of foo_get_1 to foo_1_get correct?

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks pretty good. I have some minor questions.

File: src/compiler/factory/nodeFactory.ts [hunk 12]
```diff
@@ -996,6 +1037,9 @@ namespace ts {
                 case SyntaxKind.StaticKeyword:
                     transformFlags = TransformFlags.ContainsES2015;
                     break;
+                case SyntaxKind.AccessorKeyword:
+                    transformFlags = TransformFlags.ContainsDecorators;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/factory/nodeFactory.ts:1041]: They are part of the same spec, but this is left over from before I pulled accessors out into its own PR. I will remove.

File: src/compiler/emitter.ts [hunk 13]
```diff
@@ -1191,6 +1195,10 @@ namespace ts {
             nodeIdToGeneratedName = [];
             autoGeneratedIdToGeneratedName = [];
             generatedNames = new Set();
+            formattedNameTempFlagsStack = [];
+            formattedNameTempFlags = new Map();
+            privateNameTempFlagsStack = [];
+            privateNameTempFlags = TempFlags.Auto;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/emitter.ts:1201]: Pardon?

File: src/compiler/transformers/classFields.ts [hunk 14]
```diff
@@ -133,6 +143,15 @@ namespace ts {
         const languageVersion = getEmitScriptTarget(compilerOptions);
         const useDefineForClassFields = getUseDefineForClassFields(compilerOptions);
 
+        // We need to transform class field initializers when not using Define, or when using
+        // Define and targeting < ES2022.
+        const shouldTransformInitializersUsingSet = !useDefineForClassFields || languageVersion < ScriptTarget.ES2022;
+        const shouldTransformInitializersUsingDefine = useDefineForClassFields || languageVersion < ScriptTarget.ES2022;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/transformers/classFields.ts:149]: This is a typo. Will fix and verify tests.

File: src/compiler/transformers/classFields.ts [hunk 15]
```diff
@@ -988,7 +1140,7 @@ namespace ts {
             }
             for (const member of node.members) {
                 if (!isStatic(member)) continue;
-                if (member.name && isPrivateIdentifier(member.name) && shouldTransformPrivateElementsOrClassStaticBlocks) {
+                if (member.name && (isPrivateIdentifier(member.name) || isAutoAccessorPropertyDeclaration(member)) && shouldTransformPrivateElementsOrClassStaticBlocks) {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/transformers/classFields.ts:1143]: Not necessarily. Private fields and `static {}` blocks are part of ES2022, but `accessor` fields are not. If `shouldTransformPrivateElementsOrClassStaticBlocks` is `true`, then `shouldTransformAutoAccessors` will also be `true`, not the other way around.

File: src/compiler/transformers/classFields.ts [hunk 16]
```diff
@@ -1802,99 +1976,61 @@ namespace ts {
...
-                            isValid,
-                        });
-                    }
-                }
-                else {
-                    Debug.assertNever(node, "Unknown class element type.");
-                }
+        function addPrivateIdentifierClassElementToEnvironment<T extends PropertyDeclaration | MethodDeclaration | GetAccessorDeclaration | SetAccessorDeclaration>(node: T, name: PrivateIdentifier, lex: ClassLexicalEnvironment, privateEnv: PrivateIdentifierEnvironment, isStatic: boolean, isValid: boolean, previousInfo: PrivateIdentifierInfo | undefined) {
+            if (isAutoAccessorPropertyDeclaration(node)) {
+                addPrivateIdentifierAutoAccessorPropertyDeclarationToEnvironment(node, name, lex, privateEnv, isStatic, isValid, previousInfo);
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/transformers/classFields.ts:1993]: That ends up being far less type safe since returning the function based on a narrower type doesn't narrow the type itself, so you have to have the result accept a wider type.

File: tests/baselines/reference/autoAccessorExperimentalDecorators(target=esnext).errors.txt (hunk 10 above)
Comment (rbuckton) [reply tests/baselines/reference/autoAccessorExperimentalDecorators(target=esnext).errors.txt:19]: Yes. Legacy decorators aren't supported on members with private identifiers, but will be supported for ES decorators.

File: tests/baselines/reference/privateNameDuplicateField.js (hunk 11 above)
Comment (rbuckton) [reply tests/baselines/reference/privateNameDuplicateField.js:560]: Yes. This is an intended consequence of the changes I made to allow prefixes and suffixes in name generation.

Comment (rbuckton) [reply tests/baselines/reference/privateNameDuplicateField.js:560]: However, it does look like its unused in any case.

File: tests/baselines/reference/autoAccessor7(target=es2022,usedefineforclassfields=false).js (hunk 9 above)
Comment (rbuckton) [reply tests/baselines/reference/autoAccessor7(target=es2022,usedefineforclassfields=false).js:21]: If you are using `Set` instead of `Define` and have other public fields, they all need to be moved to the constructor:

```ts
// ts
class C2 extends C1 {
  x = 1;
  accessor a = 2;
  y = 3;
}

// js, useDefineForClassFields: true
class C2 extends C1 {
  x = 1;
  #a_accessor_storage = 2;
  y = 3;
  get a() { return this.#a_accessor_storage; }
  set a(value) { this.#a_accessor_storage = value; }
}

// js, useDefineForClassFields: false
class C2 extends C1 {
  constructor() {
    super(...arguments);
    this.x = 1;
    this.#a_accessor_storage = 2;
    this.y = 3;
  }
  #a_accessor_storage;
  get a() { return this.#a_accessor_storage; }
  set a(value) { this.#a_accessor_storage = value; }
}
```

Simplifying the emit for the corner case of `useDefineForClassFields: false` when there are no other non-private fields seems like overkill.

File: tests/baselines/reference/autoAccessor6(target=esnext,usedefineforclassfields=false).js (hunk 8 above)
Comment (rbuckton) [reply tests/baselines/reference/autoAccessor6(target=esnext,usedefineforclassfields=false).js:17]: Actually, this is wrong. The field should have been included in the emit. I'll investigate.

File: tests/baselines/reference/autoAccessor5(target=es2015).js (hunk 7 above)
Comment (rbuckton) [reply tests/baselines/reference/autoAccessor5(target=es2015).js:11]: The error is that `f()` isn't a literal or unique symbol type. I can look at changing the test.

File: src/compiler/emitter.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (sandersn) [reply src/compiler/emitter.ts:1201]: sorry for the slang. I just meant that we have to write this save/restore pattern everywhere we want to emulate dynamic scope, and it would be nice if Javascript had dynamically scoped variables. They are so rarely used and easily abused that it's a not realistic language feature, though.

**Comment by rbuckton** [comment]
@sandersn I've updated based on your feedback as well as some additional cleanup of the classFields transform. Can you take another look?

**Review by sandersn — APPROVED** [review]

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by fwienber** [comment]
Also,

````typescript
class Foo {
    accessor declare foo: number;
}
````typescript

works as I would expect. Is there any reason to not support it?

**Comment by ClementValot** [comment]
Very naive question: 
Is there a reason not to implement a similar keyword for public readonly properties which would do the same, but without a setter?

**Comment by wycats** [comment]
> Is there a reason not to implement a similar keyword for public readonly properties which would do the same, but without a setter?

More generally, is there a reason that `readonly accessor` isn't supported?

**Comment by rbuckton** [comment]
> > Is there a reason not to implement a similar keyword for public readonly properties which would do the same, but without a setter?
> 
> More generally, is there a reason that `readonly accessor` isn't supported?

It's not currently supported because the accessor isn't actually read-only (a setter is still generated) and that might break user expectations. Also, support for actual read-only accessors is planned as part of https://github.com/tc39/proposal-grouped-and-auto-accessors, which is blocked on any advancement until after Decorators reaches Stage 4.

With the full auto-accessors proposal, you could accomplish a true read-only accessor using the following syntax:

```js
class C {
  accessor x { get; } = 1;
}
```

Which _would_ transform into something like this:

```js
class C {
    #x_accessor_storage = 1;
    get x() { return this.#x_accessor_storage; }
}
```

**Comment by Eghizio** [comment]
Actually having a `readonly accessor x;` would be lovely to just generate the public `getter` ❤️
